#                                                        APS - OTIMIZAÇÃO PELO VETOR GRADIENTE

Professor: Thiago Tolosa

Equipe: 
Ana Laiz Farias,
Alexandre Santarossa,
Gustavo Mendes

Nossas funções são:
$$ f(x,y)=3x^2+3xy+2y^2+x-y $$ 
$$ g(x,y)=√(x^2+y^2+2)+2x^2 e^(-y^2 )+(x-2)^2 $$ 
$$ h(x,y)=4e^(-x^2-y^2 )+3e^(-x^2-y^2+4x+6y-13)-x^2/5-y^2/9+2 $$ 


1ª Etapa (35%): Considere a função f cuja lei f(x,y) será atribuída a seu grupo.

a) Construa o gráfico de f no GeoGebra e observe que f possui um único ponto de mínimo e não possui pontos de máximo.


<img src="img/f1.png">

Observando o GeoGebra, notamos que a função é côncava para cima, significando a presença de apenas um ponto de mínimo e nenhum ponto de máximo

b) Determine o vetor gradiente de f em um ponto genérico (x,y). 

Para determinar o vetor gradiente da nossa função, utilizamos a seguinte fórmula: <p>
∇f(x,y) =( ∂f/∂x,∂f/∂y )<p>
Calculando as derivadas parciais, chegamos em:<p>
∂f/∂x=6x+3y+1 <p>
∂f/∂y=3x+4y-1 <p>
Portanto, o nosso vetor gradiente de f em um ponto genérico (x,y) é igual a:<p>
∇f(x,y)=(6x+3y+1 ,3x+4y-1)



c) Usando as ideias desenvolvidas na página anterior, elabore um código que permita determinar o ponto de mínimo da função f. Utilize um passo fixo α = 0,1 e a estimativa inicial (x₀,y₀) = (0,0). A precisão do cálculo deverá ser de 10⁻⁵. 

Usando 0,1 como passo fixo e precisão de 5 casas decimais, foi desenvolvido o seguinte código, onde é definida a função e o vetor gradiente, e após isso, entramos em um loop para calcularmos a partir do x e y iniciais (x0 e y0) os próximos valores (x1 e y2), até satisfazer a precisão, no código, cada passo está detalhado melhor com um comentário:

In [12]:
# Definindo nossa função
def f(x, y):
    return 3*x**2 + 3*x*y + 2*y**2 + x - y

# Definindo nosso vetor gradiente
def gradiente(x, y):
    return 6*x + 3*y + 1, 3*x + 4*y - 1

# Calculando o ponto de mínimo da nossa função
def ponto_minimo(passo, x0, y0, precisao):

    # Valores iniciais
    x = x0
    y = y0

    while True:

        # Definindo as componentes do valor gradiente
        grad_x, grad_y = gradiente(x, y)

        # Fórmula dada no enunciado
        x1 = x - passo * grad_x
        y1 = y - passo * grad_y

        # SE a o valor da função no ponto atual menos o ponto anterior for menor que a precisão definida, encontramos o nosso ponto de mínimo
        if abs(f(x1, y1) - f(x, y)) < precisao:
            break

        # SENÃO, atualizamos os valores de x e y para os valores calculados acima
        x = x1
        y = y1

    return x, y

ponto_minimo = ponto_minimo(0.1, 0, 0, 1e-5)
print("Ponto de mínimo:", ponto_minimo)

Ponto de mínimo: (-0.46389452400196896, 0.5961538575053097)


Assim, foi encontrado o ponto mínimo (-0.46389452400196896, 0.5961538575053097), substituindo na nossa função, encontramos o valor de -0.5333126795514063. Conferindo o resultado no SymboLab:<p>
<img src="img/symbo1.png">

Sendo x = -0.46666666666  e y = 0.6 <p>
Ou seja, nossos valores, estão bem próximos do valor do SymboLab, que provavelmente estão arredondados.


d) Repita o procedimento para os seguintes valores do passo: α = 0,15, α = 0,2, α = 0,3 e α = 0,5. 

In [19]:
# Definindo nossa função
def f(x, y):
    return 3*x**2 + 3*x*y + 2*y**2 + x - y

# Definindo nosso vetor gradiente
def gradiente(x, y):
    return 6*x + 3*y + 1, 3*x + 4*y - 1

# Calculando o ponto de mínimo da nossa função
def ponto_minimo(passo, x0, y0, precisao):

    # Valores iniciais
    x = x0
    y = y0

    while True:

        # Definindo as componentes do valor gradiente
        grad_x, grad_y = gradiente(x, y)

        # Fórmula dada no enunciado
        x1 = x - passo * grad_x
        y1 = y - passo * grad_y

        # SE a o valor da função no ponto atual menos o ponto anterior for menor que a precisão definida, encontramos o nosso ponto de mínimo
        if abs(f(x1, y1) - f(x, y)) < precisao:
            break

        # SENÃO, atualizamos os valores de x e y para os valores calculados acima
        x = x1
        y = y1

    return x, y

ponto_minimo_015 = ponto_minimo(0.15, 0, 0, 1e-5)
print("Ponto de mínimo com passo 0.15:", ponto_minimo_015)
ponto_minimo_02 = ponto_minimo(0.2, 0, 0, 1e-5)
print("Ponto de mínimo com passo 0.2:", ponto_minimo_02)

Ponto de mínimo com passo 0.15: (-0.4641161580836166, 0.5964613583695217)
Ponto de mínimo com passo 0.2: (-0.4647551999999999, 0.5975424)


Para os valores de passo de 0,15 e 0,2, conseguimos utilizar o mesmo código do item c), porém, ao partimos para os valores de 0,3 e 0,5 nos deparamos com um erro Overflow no código, significando que os valores estavam ficando muito grandes para serem processados pelo computador. Deste modo, o único método foi limitar o código para repetir a conta apenas 200 vezes, afetando a precisão do resultado, mas ainda assim, conseguimos encontrar o ponto mínimo, porém, com uma precisão bem menor.

In [24]:
# Definindo nossa função
def f(x, y):
    return 3*x**2 + 3*x*y + 2*y**2 + x - y

# Definindo nosso vetor gradiente
def gradiente(x, y):
    return 6*x + 3*y + 1, 3*x + 4*y - 1

# Calculando o ponto de mínimo da nossa função
def ponto_minimo(passo, x0, y0, precisao):

    # Valores iniciais
    x = x0
    y = y0
    contador = 0
    menorX = [0]
    menorY = [0]
    while True:

        # Definindo as componentes do valor gradiente
        grad_x, grad_y = gradiente(x, y)

        # Fórmula dada no enunciado
        x1 = x - passo * grad_x
        y1 = y - passo * grad_y

        # SE a o valor da função no ponto atual menos o ponto anterior for menor que a precisão definida, encontramos o nosso ponto de mínimo
        if abs(f(x1, y1) - f(x, y)) < precisao or contador > 200:
            break

        # SENÃO, atualizamos os valores de x e y para os valores calculados acima
        x = x1
        y = y1
        contador += 1

        # Analisamos o valor da função no ponto atual e no ponto de mínimo anterior e atualizamos o ponto de mínimo caso necessário
        if f(x1, y1) < f(menorX[0], menorY[0]):
            menorX[0] = x1
            menorY[0] = y1
    
    print(f"Valor da função no ponto de mínimo com passo {passo} :", f(menorX[0], menorY[0]))

    # Retornamos o ponto de mínimo
    return menorX[0], menorY[0]


# COLOQUE O PASSO DESEJADO NA ENTRADA DA FUNÇÃO
ponto_minimo_03 = ponto_minimo(0.3, 0, 0, 1e-5)
print("Ponto de mínimo com passo 0.3:", ponto_minimo_03)
ponto_minimo_05 = ponto_minimo(0.5, 0, 0, 1e-5)
print("Ponto de mínimo com passo 0.5:", ponto_minimo_05)


Valor da função no ponto de mínimo com passo 0.3 : -0.49995
Ponto de mínimo com passo 0.3: (-0.495, 0.495)
Valor da função no ponto de mínimo com passo 0.5 : -0.5
Ponto de mínimo com passo 0.5: (-0.5, 0.5)


2ª Etapa (20%): Utilize o seu código para determinar os pontos de mínimo da função g, cuja lei g(x,y) será atribuída a seu grupo. A função g possui dois pontos de mínimo. <p>
No relatório, deverá aparecer: <p>
    • Os dois pontos de mínimo obtidos com seu código; <p>
    • O que teve que ser modificado no programa para obter o segundo ponto de mínimo; <p>
    • O comportamento do processo de convergência para diferentes valores do passo. <p>


In [2]:
import numpy as np
import math
def g(x,y):
    g = math.sqrt(x**2+y**2+2)+2*(x**2)*math.exp(-y**2)+(x-2)**2
    return g
def gradiente_de_g(x, y):
    df_dx = (4*x*math.exp(-y**2) + x*((x**2+y**2+2)**(-0.5)) + 2*(x-2))
    df_dy = (-4*(x**2)*y*math.exp(-y**2) + y*((x**2 + y**2 + 2)**(-0.5)))
    return np.array([df_dx, df_dy])

def grad_descend(x0, y0, alpha):
    x, y = x0, y0
    #conta o numero de iterações para chegar no mínimo
    cont = 0
    while True:
        gradiente = gradiente_de_g(x, y)
        if np.linalg.norm(gradiente) <= 1e-5:
            break
        x -= alpha*gradiente[0]
        y -= alpha*gradiente[1]
        cont += 1
    return (x, y),cont

# primeiro ponto mínimo de g
x1 = -2
y1 = -2

ponto1 , cont1 = grad_descend(x1, y1,0.1)

print("Primeiro ponto mínimo: ({:.5f}, {:.5f}) e o numero de iterações pra chegar no mínimo foi de {}".format(ponto1[0],ponto1[1],cont1))

# Procurando o segundo ponto mínimo da função g
x2 = 2
y2 = 2

ponto2 , cont2 = grad_descend(x2, y2,0.1)

print("Segundo ponto mínimo: ({:.5f}, {:.5f}) e o numero de iterações pra chegar no mínimo foi de {}".format(ponto2[0],ponto2[1],cont2))

# para diferentes passos: 
passo_values = [0.15, 0.2, 0.3, 0.5]
for i in range(len(passo_values)):
    x1, y1 = 2, 2
    ponto1 , cont1 = grad_descend(x1, y1,passo_values[i])

    print("Primeiro ponto mínimo para alpha = {}: ({:.5f}, {:.5f}) e o numero de iterações pra chegar no mínimo foi de {}".format(passo_values[i],ponto1[0],ponto1[1],cont1))


for i in range(len(passo_values)):
    x2, y2 = -2, -2
    ponto2 , cont2 = grad_descend(x2, y2,passo_values[i])

    print("Segundo ponto mínimo para alpha = {}: ({:.5f}, {:.5f}) e o numero de iterações pra chegar no mínimo foi de {}".format(passo_values[i],ponto2[0],ponto2[1],cont2))



Primeiro ponto mínimo: (1.60476, -1.83533) e o numero de iteções pra chegar no mínimo foi de 84
Segundo ponto mínimo: (1.60477, 1.83534) e o numero de iteções pra chegar no mínimo foi de 74
Primeiro ponto mínimo para alpha = 0.15: (1.60477, 1.83534) e o numero de iteções pra chegar no mínimo foi de 48
Primeiro ponto mínimo para alpha = 0.2: (1.60477, 1.83534) e o numero de iteções pra chegar no mínimo foi de 34
Primeiro ponto mínimo para alpha = 0.3: (1.60477, 1.83533) e o numero de iteções pra chegar no mínimo foi de 21
Primeiro ponto mínimo para alpha = 0.5: (1.60477, 1.83533) e o numero de iteções pra chegar no mínimo foi de 22
Segundo ponto mínimo para alpha = 0.15: (1.60476, -1.83533) e o numero de iteções pra chegar no mínimo foi de 54
Segundo ponto mínimo para alpha = 0.2: (1.60476, -1.83533) e o numero de iteções pra chegar no mínimo foi de 38
Segundo ponto mínimo para alpha = 0.3: (1.60476, -1.83533) e o numero de iteções pra chegar no mínimo foi de 22
Segundo ponto mínimo par

3ª Etapa (20%): Utilize o seu código para determinar os dois pontos de máximo da função ℎ, cuja lei h(x,y) será atribuída a seu grupo. No relatório, indique os dois pontos de máximo obtidos com seu código. 

In [3]:
import numpy as np
import math

def h(x,y):
    h = 4*math.exp(-x**2-y**2 )+3*math.exp(-x**2-y**2+4*x+6*y-13)-((x**2)/5)-((y**2)/9) + 2
    return h
def gradiente_de_h(x, y):
    df_dx = (-6*(x-2)*(math.exp(-x**2 + 4*x - y**2 + 6*y - 13)) - (8*x)*(math.exp(-x**2 - y**2)) - ((2*x)/5))
    df_dy = (-6*(y-3)*(math.exp(-x**2 + 4*x - y**2 + 6*y - 13)) - (8*y)*(math.exp(-x**2 - y**2)) - ((2*y)/9))
    return np.array([df_dx, df_dy])

def grad_descend(x0, y0, alpha):
    x, y = x0, y0
    #conta o numero de iterações para chegar no mínimo
    cont = 0
    while True:
        gradiente = gradiente_de_h(x, y)
        if np.linalg.norm(gradiente) <= 1e-5:
            break
        x += alpha*gradiente[0]
        y += alpha*gradiente[1]
        cont += 1
    return (x, y),cont

# primeiro ponto mínimo de f
x1 = -2
y1 = -2

ponto1 , cont1 = grad_descend(x1, y1,0.1)

print("Primeiro ponto máximo: ({:.5f}, {:.5f}) e o numero de iteções pra chegar no mínimo foi de {}".format(ponto1[0],ponto1[1],cont1))

# Procurando o segundo ponto mínimo da função f
x2 = 2
y2 = 2

ponto2 , cont2 = grad_descend(x2, y2,0.1)

print("Segundo ponto máximo: ({:.5f}, {:.5f}) e o numero de iteções pra chegar no mínimo foi de {}".format(ponto2[0],ponto2[1],cont2))

# para diferentes passos: 
passo_values = [0.15, 0.2, 0.3, 0.5]
for i in range(len(passo_values)):
    x1, y1 = 2, 2
    ponto1 , cont1 = grad_descend(x1, y1,passo_values[i])

    print("Primeiro ponto máximo para alpha = {}: ({:.5f}, {:.5f}) e o numero de iteções pra chegar no mínimo foi de {}".format(passo_values[i],ponto1[0],ponto1[1],cont1))


for i in range(len(passo_values)):
    x2, y2 = -2, -2
    ponto2 , cont2 = grad_descend(x2, y2,passo_values[i])

    print("Segundo ponto máximo para alpha = {}: ({:.5f}, {:.5f}) e o numero de iteções pra chegar no mínimo foi de {}".format(passo_values[i],ponto2[0],ponto2[1],cont2))




Primeiro ponto máximo: (0.00000, 0.00000) e o numero de iteções pra chegar no mínimo foi de 27
Segundo ponto máximo: (1.87159, 2.88983) e o numero de iteções pra chegar no mínimo foi de 18
Primeiro ponto máximo para alpha = 0.15: (1.87159, 2.88983) e o numero de iteções pra chegar no mínimo foi de 9
Primeiro ponto máximo para alpha = 0.2: (1.87159, 2.88983) e o numero de iteções pra chegar no mínimo foi de 9
Primeiro ponto máximo para alpha = 0.3: (1.87159, 2.88983) e o numero de iteções pra chegar no mínimo foi de 65


4ª Etapa (25%): Faça uma pesquisa sobre a utilização de passo variável e implemente o método no seu código para calcular o ponto de mínimo da função f. Escreva um breve relato comparando o número de iterações necessárias até a convergência nos dois casos.